Mathlib/Algebra/BigOperators/Expect.lean

```diff
@@ -0,0 +1,403 @@
+domain of the average, rather than being a non-dependent function. -/
+lemma expect_bij (i : ∀ a ∈ s, κ) (hi : ∀ a ha, i a ha ∈ t) (h : ∀ a ha, f a = g (i a ha))
+    (i_inj : ∀ a₁ ha₁ a₂ ha₂, i a₁ ha₁ = i a₂ ha₂ → a₁ = a₂)
+    (i_surj : ∀ b ∈ t, ∃ a ha, i a ha = b) : 𝔼 i ∈ s, f i = 𝔼 i ∈ t, g i := by
+  simp_rw [expect, card_bij i hi i_inj i_surj, sum_bij i hi i_inj i_surj h]
+
+/-- Reorder an average.
+
+The difference with `Finset.expect_bij` is that the bijection is specified with an inverse, rather
+than as a surjective injection.
+
+The difference with `Finset.expect_nbij'` is that the bijection and its inverse are allowed to use
+membership of the domains of the averages, rather than being non-dependent functions. -/
+lemma expect_bij' (i : ∀ a ∈ s, κ) (j : ∀ a ∈ t, ι) (hi : ∀ a ha, i a ha ∈ t)
+    (hj : ∀ a ha, j a ha ∈ s) (left_inv : ∀ a ha, j (i a ha) (hi a ha) = a)
+    (right_inv : ∀ a ha, i (j a ha) (hj a ha) = a) (h : ∀ a ha, f a = g (i a ha)) :
+    𝔼 i ∈ s, f i = 𝔼 i ∈ t, g i := by
+  simp_rw [expect, card_bij' i j hi hj left_inv right_inv, sum_bij' i j hi hj left_inv right_inv h]
+
+/-- Reorder an average.
+
+The difference with `Finset.expect_nbij'` is that the bijection is specified as a surjective
+injection, rather than by an inverse function.
+
+The difference with `Finset.expect_bij` is that the bijection is a non-dependent function, rather
+than being allowed to use membership of the domain of the average. -/
+lemma expect_nbij (i : ι → κ) (hi : ∀ a ∈ s, i a ∈ t) (h : ∀ a ∈ s, f a = g (i a))
+    (i_inj : (s : Set ι).InjOn i) (i_surj : (s : Set ι).SurjOn i t) :
+    𝔼 i ∈ s, f i = 𝔼 i ∈ t, g i := by
+  simp_rw [expect, card_nbij i hi i_inj i_surj, sum_nbij i hi i_inj i_surj h]
+
+/-- Reorder an average.
+
+The difference with `Finset.expect_nbij` is that the bijection is specified with an inverse, rather
+than as a surjective injection.
+
+The difference with `Finset.expect_bij'` is that the bijection and its inverse are non-dependent
+functions, rather than being allowed to use membership of the domains of the averages.
+
+The difference with `Finset.expect_equiv` is that bijectivity is only required to hold on the
+domains of the averages, rather than on the entire types. -/
+lemma expect_nbij' (i : ι → κ) (j : κ → ι) (hi : ∀ a ∈ s, i a ∈ t) (hj : ∀ a ∈ t, j a ∈ s)
+    (left_inv : ∀ a ∈ s, j (i a) = a) (right_inv : ∀ a ∈ t, i (j a) = a)
+    (h : ∀ a ∈ s, f a = g (i a)) : 𝔼 i ∈ s, f i = 𝔼 i ∈ t, g i := by
+  simp_rw [expect, card_nbij' i j hi hj left_inv right_inv,
+    sum_nbij' i j hi hj left_inv right_inv h]
+
+/-- `Finset.expect_equiv` is a specialization of `Finset.expect_bij` that automatically fills in
+most arguments. -/
+lemma expect_equiv (e : ι ≃ κ) (hst : ∀ i, i ∈ s ↔ e i ∈ t) (hfg : ∀ i ∈ s, f i = g (e i)) :
+    𝔼 i ∈ s, f i = 𝔼 i ∈ t, g i := by simp_rw [expect, card_equiv e hst, sum_equiv e hst hfg]
+
+lemma expect_product (f : ι × κ → M) : 𝔼 x ∈ s ×ˢ t, f x = 𝔼 i ∈ s, 𝔼 j ∈ t, f (i, j) := by
```

Comment:
Let's add a comment about the rewrite direction of these, like we have in the finset versions of similar lemmas